[FEATURE] Improve error message when a node with an incorrectly configured certificate attempts to connect

**Problem**

When setting up a cluster, if a node tries to join with a certificate that is not configured correctly in [plugins.security.nodes_dn](https://opensearch.org/docs/latest/security/configuration/tls/#configuring-node-certificates), OpenSearch provides an error message - "Transport client authentication no longer supported". 

```
[2024-07-26T03:02:58,137][WARN ][o.o.d.HandshakingTransportAddressConnector] [os01] handshake failed for [connectToRemoteMasterNode[172.23.0.2:9300]]
opensearch-docker-test-os01-1    | org.opensearch.transport.RemoteTransportException: [os02][172.23.0.2:9300][internal:transport/handshake]
opensearch-docker-test-os01-1    | Caused by: org.opensearch.OpenSearchException: Transport client authentication no longer supported.
```

This is a misleading error - under this circumstance, the problem has nothing to do with the transport client.

**Context of the error message**

My understanding of this error message is as follows.

OpenSearch clusters have two different ports:
- The REST API port, usually port 9200
- The internode communication port (aka 'transport' port), usually port 9300
 
In the early days of Elasticsearch, there was a popular application client, the transport client, that connected on the internode port. This old method of connecting has been deprecated since ES 7.0 in 2019. It was removed in OpenSearch 2.0 (https://github.com/opensearch-project/security/pull/1701). Modern application clients connect on the REST API port.
 
From the perspective of OpenSearch, an incoming connection on port 9300 could be coming from a node, or it could be coming from a transport client.

There is some discussion of this in issue https://github.com/opensearch-project/security/issues/4401 and the linked forum posts https://forum.opensearch.org/t/transport-client-authentication-no-longer-supported-error-when-deploying-cluster-with-security-plugin-enabled/9814/5 / https://forum.opensearch.org/t/transport-client-authentication-no-longer-supported-error-while-implementing-third-party-ca-cert-for-transport-layer/19694 . 

**Proposed solution**

I would like to make this error message clearer.

If we could distinguish between a node joining and a transport client connecting, and provide different error messages for each, that would be ideal.

If it is not easy to tell the difference from the server side, we could still add more information to the error message - something like:

"A connection was attempted on this cluster's internode communication port, but it was refused. If you are attempting to join a node to this cluster, please confirm the Distinguished Name of the new node's certificate is configured in this node's plugins.security.nodes_dn setting. If you are attempting to connect to this cluster with a transport client, please note that transport clients are no longer supported."

It would be nice to also mention the DN that was rejected - for example:

"The distinguished name of the node certificate provided was: 'CN=node.other.com,OU=SSL,O=Test,L=Test,C=DE'"

This would help the user work out what needs to be configured in plugins.security.nodes_dn .

--- a/src/main/java/org/opensearch/security/ssl/util/ExceptionUtils.java
+++ b/src/main/java/org/opensearch/security/ssl/util/ExceptionUtils.java
@@ -77,8 +77,12 @@ public class ExceptionUtils {
     }
 
     public static OpenSearchException clusterWrongNodeCertConfigException(String sslPrincipal) {
-        return new OpenSearchException("Node presenting certificate with SSL Principal {" + sslPrincipal + "} could" +
-                " not securely connect to the cluster. Please ensure the principal is correct and present in the" +
-                " nodes_dn list.");
+        return new OpenSearchException(
+            "Node presenting certificate with SSL Principal {"
+                + sslPrincipal
+                + "} could"
+                + " not securely connect to the cluster. Please ensure the principal is correct and present in the"
+                + " nodes_dn list."
+        );
     }
 }

--- a/src/main/java/org/opensearch/security/ssl/util/ExceptionUtils.java
+++ b/src/main/java/org/opensearch/security/ssl/util/ExceptionUtils.java
@@ -76,7 +76,7 @@ public class ExceptionUtils {
         return new OpenSearchException("An error occurred during the creation of Jwk: {}", cause, cause.getMessage());
     }
 
-    public static OpenSearchException createTransportClientNoLongerSupportedException() {
-        return new OpenSearchException("Transport client authentication no longer supported.");
+    public static OpenSearchException clusterWrongNodeCertConfigException() {
+        return new OpenSearchException("Node certificate configuration is wrong or certificate is invalid.");
     }
 }

--- a/src/main/java/org/opensearch/security/ssl/util/ExceptionUtils.java
+++ b/src/main/java/org/opensearch/security/ssl/util/ExceptionUtils.java
@@ -76,7 +76,9 @@ public class ExceptionUtils {
         return new OpenSearchException("An error occurred during the creation of Jwk: {}", cause, cause.getMessage());
     }
 
-    public static OpenSearchException clusterWrongNodeCertConfigException() {
-        return new OpenSearchException("Node certificate configuration is wrong or certificate is invalid.");
+    public static OpenSearchException clusterWrongNodeCertConfigException(String sslPrincipal) {
+        return new OpenSearchException("Node presenting certificate with SSL Principal {" + sslPrincipal + "} could" +
+                " not securely connect to the cluster. Please ensure the principal is correct and present in the" +
+                " nodes_dn list.");
     }
 }

--- a/src/main/java/org/opensearch/security/transport/SecurityRequestHandler.java
+++ b/src/main/java/org/opensearch/security/transport/SecurityRequestHandler.java
@@ -290,7 +290,7 @@ public class SecurityRequestHandler<T extends TransportRequest> extends Security
                     || HeaderHelper.isTrustedClusterRequest(getThreadContext())
                     || HeaderHelper.isExtensionRequest(getThreadContext()))) {
                     // CS-ENFORCE-SINGLE
-                    final OpenSearchException exception = ExceptionUtils.createTransportClientNoLongerSupportedException();
+                    final OpenSearchException exception = ExceptionUtils.clusterWrongNodeCertConfigException();
                     log.error(exception.toString());
                     transportChannel.sendResponse(exception);
                     return;

--- a/src/main/java/org/opensearch/security/transport/SecurityRequestHandler.java
+++ b/src/main/java/org/opensearch/security/transport/SecurityRequestHandler.java
@@ -290,7 +290,7 @@ public class SecurityRequestHandler<T extends TransportRequest> extends Security
                     || HeaderHelper.isTrustedClusterRequest(getThreadContext())
                     || HeaderHelper.isExtensionRequest(getThreadContext()))) {
                     // CS-ENFORCE-SINGLE
-                    final OpenSearchException exception = ExceptionUtils.clusterWrongNodeCertConfigException();
+                    final OpenSearchException exception = ExceptionUtils.clusterWrongNodeCertConfigException(principal);
                     log.error(exception.toString());
                     transportChannel.sendResponse(exception);
                     return;

--- a/src/test/java/org/opensearch/security/ccstest/CrossClusterSearchTests.java
+++ b/src/test/java/org/opensearch/security/ccstest/CrossClusterSearchTests.java
@@ -1355,7 +1355,9 @@ public class CrossClusterSearchTests extends AbstractSecurityUnitTest {
         String uri = "cross_cluster_two:twitter/_search?pretty";
         HttpResponse ccs = rh1.executeGetRequest(uri, encodeBasicHeader("twitter", "nagilum"));
         assertThat(ccs.getStatusCode(), equalTo(HttpStatus.SC_INTERNAL_SERVER_ERROR));
-        assertThat(ccs.getBody(), containsString("Node certificate configuration is wrong or certificate is invalid."));
+        assertThat(ccs.getBody(), containsString("Node presenting certificate with SSL Principal " +
+                "{CN=node-0.example.com,OU=SSL,O=Test,L=Test,C=DE} could not securely connect to the cluster. Please" +
+                " ensure the principal is correct and present in the nodes_dn list."));
     }
 
     @Test

--- a/src/test/java/org/opensearch/security/ccstest/CrossClusterSearchTests.java
+++ b/src/test/java/org/opensearch/security/ccstest/CrossClusterSearchTests.java
@@ -1355,9 +1355,14 @@ public class CrossClusterSearchTests extends AbstractSecurityUnitTest {
         String uri = "cross_cluster_two:twitter/_search?pretty";
         HttpResponse ccs = rh1.executeGetRequest(uri, encodeBasicHeader("twitter", "nagilum"));
         assertThat(ccs.getStatusCode(), equalTo(HttpStatus.SC_INTERNAL_SERVER_ERROR));
-        assertThat(ccs.getBody(), containsString("Node presenting certificate with SSL Principal " +
-                "{CN=node-0.example.com,OU=SSL,O=Test,L=Test,C=DE} could not securely connect to the cluster. Please" +
-                " ensure the principal is correct and present in the nodes_dn list."));
+        assertThat(
+            ccs.getBody(),
+            containsString(
+                "Node presenting certificate with SSL Principal "
+                    + "{CN=node-0.example.com,OU=SSL,O=Test,L=Test,C=DE} could not securely connect to the cluster. Please"
+                    + " ensure the principal is correct and present in the nodes_dn list."
+            )
+        );
     }
 
     @Test

--- a/src/test/java/org/opensearch/security/ccstest/CrossClusterSearchTests.java
+++ b/src/test/java/org/opensearch/security/ccstest/CrossClusterSearchTests.java
@@ -1355,7 +1355,7 @@ public class CrossClusterSearchTests extends AbstractSecurityUnitTest {
         String uri = "cross_cluster_two:twitter/_search?pretty";
         HttpResponse ccs = rh1.executeGetRequest(uri, encodeBasicHeader("twitter", "nagilum"));
         assertThat(ccs.getStatusCode(), equalTo(HttpStatus.SC_INTERNAL_SERVER_ERROR));
-        assertThat(ccs.getBody(), containsString("Transport client authentication no longer supported"));
+        assertThat(ccs.getBody(), containsString("Node certificate configuration is wrong or certificate is invalid."));
     }
 
     @Test

